### PR TITLE
Abstract commonly used functionality from `EntitySource`.

### LINKED ISSUE #975: Make `EntitySource` completely static
After refinements related to #969, the `DataSource` that is required for instanciating `EntitySource` is effectively only used in `EntitySource#buildEntityData`. Since all other methods are static or build upon this method, one can get confused easily here: When invoking one of these methods, does the output result from the input parameters, the field `EntitySource.dataSource`, or both?

Thus, I'm suggesting to make `EntitySource` completely static. Where data is needed from `DataSource`, either this object or a list of field-to-attribute sets can be passed as a parameter.

One can also discuss whether to also apply this transformation to subclasses of `EntitySource`, although this might be a bit more complex.

### LINKED ISSUE #981: Abstract commonly used functionality from `EntitySource`
Especially with #926 it became apparent that some sources re-implement functionality that already exists in `EntitySource`.
This functionality includes (but might not be limited to) methods 

```java
public abstract void validate() throws ValidationException;
<C extends UniqueEntity> Try<Void, ValidationException> validate(Class<? extends C>, SourceValidator<C>)
```
, which could be reused by classes
- `*WeatherSource` (`CsvWeatherSource` might be a bit more difficult)
- `*IdCoordinateSource`
- `*TimeSeriesSource`

where `*` can be multiple of `Sql`, `Csv`, `Couchbase`, `InfluxDb`.

Probably the methods above have to be adapted to abstract the functionality properly.

In the end, we'll probably have `AssetEntitySource`, which takes over most of the current `EntitySource` implementation, and a new `EntitySource` which will be extended by `AssetEntitySource` and by the classes mentioned above.

### PR DESCRIPTION
**Pull request by staudtMarius**
Resolves #975
Resolves #981

### CODE REVIEW

File: src/main/java/edu/ie3/datamodel/io/source/GraphicSource.java [hunk 1]
```diff
@@ -118,15 +120,18 @@ public Set<NodeGraphicInput> getNodeGraphicInput() throws SourceException {
...
+            nodeGraphicInputFactory,
+            enrich(NODE, nodes, NodeGraphicInputEntityData::new))
+        .collect(toSet());
   }
 
   /**
    * If the set of {@link LineInput} entities is not exhaustive for all available {@link
    * LineGraphicInput} entities or if an error during the building process occurs a {@link
-   * SourceException} is thrown, else all entities that have been able to be built are returned.
+   * SourceException} is thrown, else all entities that have been able to be build are returned.
```
Comment (sebastian-peter) [inline src/main/java/edu/ie3/datamodel/io/source/GraphicSource.java:134]: This was correct before, we need the past participle of "build" here

File: src/main/java/edu/ie3/datamodel/io/source/SystemParticipantSource.java [hunk 2]
```diff
@@ -680,154 +724,29 @@ public Set<HpInput> getHeatPumps(
...
-                nodeAssetInputEntityDataTry.flatMap(
-                    nodeAssetInputEntityData ->
-                        optionallyEnrichEntityData(
-                            nodeAssetInputEntityData,
-                            SystemParticipantInputEntityFactory.EM,
-                            emUnits,
-                            null,
-                            SystemParticipantEntityData::new)));
+  private static <T extends SystemParticipantTypeInput, D extends SystemParticipantEntityData>
+      TryFunction<D, SystemParticipantTypedEntityData<T>> enrich(Map<UUID, T> types) {
```
Comment (sebastian-peter) [inline src/main/java/edu/ie3/datamodel/io/source/SystemParticipantSource.java:747]: Same here as above, please rename function

File: src/main/java/edu/ie3/datamodel/io/source/csv/CsvIdCoordinateSource.java [hunk 3]
```diff
@@ -53,11 +53,21 @@ public CsvIdCoordinateSource(IdCoordinateFactory factory, CsvDataSource dataSour
     this.factory = factory;
     this.dataSource = dataSource;
 
+    // validating
+    Try.ofVoid(this::validate, ValidationException.class)
+        .transformF(SourceException::new)
+        .getOrThrow();
```
Comment (sebastian-peter) [inline src/main/java/edu/ie3/datamodel/io/source/csv/CsvIdCoordinateSource.java:59]: So here we start validating within the constructor, but in weather sources you remove the validation from constructor. Is there a reason for that?

File: src/main/java/edu/ie3/datamodel/io/source/sql/SqlIdCoordinateSource.java [hunk 4]
```diff
@@ -58,19 +56,6 @@ public SqlIdCoordinateSource(
...
-            fieldsOpt ->
-                fieldsOpt
-                    .map(
-                        fields ->
-                            factory
-                                .validate(fields, IdCoordinateInput.class)
-                                .transformF(SourceException::new))
-                    .orElse(Try.Success.empty()))
-        .getOrThrow();
-
```
Comment (sebastian-peter) [inline src/main/java/edu/ie3/datamodel/io/source/sql/SqlIdCoordinateSource.java:73]: ... adding validation to the equivalent csv source, removing it here?

File: src/test/groovy/edu/ie3/datamodel/io/source/csv/CsvSystemParticipantSourceTest.groovy [hunk 5]
```diff
@@ -159,7 +157,6 @@ class CsvSystemParticipantSourceTest extends Specification implements CsvTestDat
 
     where:
     operators               | types               | thermalBuses              || resultingSize | resultingSet
-    []                      | [sptd.hpInput.type] | [sptd.hpInput.thermalBus] || 1             | [new HpInput(sptd.hpInput.uuid, sptd.hpInput.id, OperatorInput.NO_OPERATOR_ASSIGNED, sptd.hpInput.operationTime, sptd.hpInput.node, sptd.hpInput.thermalBus, sptd.hpInput.qCharacteristics, sptd.emInput, sptd.hpInput.type)]
```
Comment (sebastian-peter) [inline src/test/groovy/edu/ie3/datamodel/io/source/csv/CsvSystemParticipantSourceTest.groovy:162]: I don't agree with just removing the successful cases completely. What is the problem here?

**Review by sebastian-peter — CHANGES_REQUESTED** [review]
I like your implementation of the issue a lot. This is more than I expected! I just have some minor remarks.

Additionally, please also adapt `EntitySourceClassDiagram.puml`. If you think that it makes sense, you can leave out protected and private methods there from now on.

File: src/test/groovy/edu/ie3/datamodel/io/source/csv/CsvSystemParticipantSourceTest.groovy (hunk 5 above)
Comment (staudtMarius) [reply src/test/groovy/edu/ie3/datamodel/io/source/csv/CsvSystemParticipantSourceTest.groovy:162]: In theses test we are explicitly testing cases that are leading to failures.

Comment (sebastian-peter) [reply src/test/groovy/edu/ie3/datamodel/io/source/csv/CsvSystemParticipantSourceTest.groovy:162]: Ok, I see. `resultingSize` and `resultingSet` is not used anymore and can be removed then

File: src/test/groovy/edu/ie3/datamodel/io/source/csv/CsvSystemParticipantSourceTest.groovy [hunk 6]
```diff
@@ -283,7 +276,6 @@ class CsvSystemParticipantSourceTest extends Specification implements CsvTestDat
 
     where:
     operators               | types               || resultingSet
-    []                      | [sptd.bmInput.type] || [new BmInput(sptd.bmInput.uuid, sptd.bmInput.id, OperatorInput.NO_OPERATOR_ASSIGNED, sptd.bmInput.operationTime, sptd.bmInput.node, sptd.bmInput.qCharacteristics, sptd.emInput, sptd.bmInput.type, sptd.bmInput.marketReaction, sptd.bmInput.costControlled, sptd.bmInput.feedInTariff)]
     [sptd.bmInput.operator] | []                  || []
     []                      | []                  || []
     []                      | []                  || []
```
Comment (sebastian-peter) [inline src/test/groovy/edu/ie3/datamodel/io/source/csv/CsvSystemParticipantSourceTest.groovy:281]: Also this is just duplicated...

File: src/main/java/edu/ie3/datamodel/utils/Try.java
Comment (sebastian-peter) [inline src/main/java/edu/ie3/datamodel/utils/Try.java:1]: Ok, something I just noticed. Could you take care of testing the new methods in `TryTest`? Thanks!

**Review by sebastian-peter — CHANGES_REQUESTED** [review]

**Review by sebastian-peter — APPROVED** [review]
Thanks!

**Pull request merged**